File: frontend/src/services/terminal.ts
import { Events } from "@wailsio/runtime";
import { TerminalService } from "../../bindings/pi-desk/internal/appservice";
import type { TerminalState } from "../../bindings/pi-desk/internal/domain";

export interface TerminalEvent {
  threadId: string;
  type: "output" | "error" | "exit";
  generation?: number;
  sequence: number;
  dataB64?: string;
  exitCode?: number;
  error?: string;
}

export type TerminalWorkspaceReference = string | { workspaceId: string };

export const terminalService = {
  start(threadId: string, workspace: TerminalWorkspaceReference, columns: number, rows: number): Promise<TerminalState> {
    const reference = typeof workspace === "string" ? { workspacePath: workspace } : workspace;
    return TerminalService.Start({ threadId, ...reference, columns, rows });
  },
  snapshot(threadId: string, workspaceId?: string): Promise<TerminalState> {
    return TerminalService.Snapshot({ threadId, workspaceId });
  },
  write(threadId: string, data: string): Promise<void> {
    return TerminalService.Write({ threadId, data });
  },
  resize(threadId: string, columns: number, rows: number): Promise<void> {
    return TerminalService.Resize({ threadId, columns, rows });
  },
  stop(threadId: string): Promise<void> {
    return TerminalService.Stop({ threadId });
  },
};

export function onTerminalEvent(callback: (event: TerminalEvent) => void): () => void {
  return Events.On("terminal:event", (event) => callback(event.data as TerminalEvent));
}

File: internal/domain/terminal.go
package domain

type StartTerminalRequest struct {
	ThreadID      string `json:"threadId"`
	WorkspaceID   string `json:"workspaceId,omitempty"`
	WorkspacePath string `json:"workspacePath,omitempty"`
	Columns       int    `json:"columns"`
	Rows          int    `json:"rows"`
}

type TerminalRequest struct {
	ThreadID    string `json:"threadId"`
	WorkspaceID string `json:"workspaceId,omitempty"`
}

type TerminalWriteRequest struct {
	ThreadID string `json:"threadId"`
	Data     string `json:"data"`
}

type TerminalResizeRequest struct {
	ThreadID string `json:"threadId"`
	Columns  int    `json:"columns"`
	Rows     int    `json:"rows"`
}

type TerminalState struct {
	ThreadID   string `json:"threadId"`
	CWD        string `json:"cwd,omitempty"`
	Shell      string `json:"shell,omitempty"`
	Running    bool   `json:"running"`
	Generation uint64 `json:"generation,omitempty"`
	Sequence   uint64 `json:"sequence"`
	OutputB64  string `json:"outputB64,omitempty"`
}

type TerminalEvent struct {
	ThreadID   string `json:"threadId"`
	Type       string `json:"type"`
	Generation uint64 `json:"generation,omitempty"`
	Sequence   uint64 `json:"sequence"`
	DataB64    string `json:"dataB64,omitempty"`
	ExitCode   int    `json:"exitCode,omitempty"`
	Error      string `json:"error,omitempty"`
}
